### PR TITLE
[v11] Simplify Component System Usage

### LINKED ISSUE #12957: [FR]: Update Component model to support registration from pure Swift
### Description

Update Component model to support registration from pure Swift without `@objc` or `class` requirements and `struct` support.

Note that it is no longer a requirement for FirebaseCore not to know about its clients now that we have uni-versioning and SPM.

### API Proposal

_No response_

### Firebase Product(s)

Core

### PR DESCRIPTION
**Pull request by paulb777**
A Component class is not needed for products that don't need to do early initialization and if all of their dependencies are Eager components.

Only registerLibrary needs to run for userAgent configuration.  This is done by checking for the existence of an `objc` class in FIRApp.m. If one doesn't exist at all a singleton empty `@objc` class is added. See the example in VertexAI.swift.

Fix a Storage unit test that was dependent on a race condition to pass. (URL overrides the bucket name when it pre-existed)

Fix #12957

TODO: Investigate deleting the `FIRDependency` class which seems to be ignored.

#no-changelog

### CODE REVIEW

File: FirebaseVertexAI/Tests/Unit/VertexComponentTests.swift [hunk 1]
```diff
@@ -1,115 +0,0 @@
...
-  /// Tests that a vertex instance can be created properly by the VertexAIComponent.
-  func testVertexInstanceCreation() throws {
-    let app = try XCTUnwrap(VertexComponentTests.app)
-    let component = VertexAIComponent(app: app)
-    let vertex = component.vertexAI("my-location")
-    XCTAssertNotNil(vertex)
-  }
-
-  /// Tests that the component container caches instances of VertexAIComponent.
-  func testMultipleComponentInstancesCreated() throws {
```
[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseVertexAI/Tests/Unit/VertexComponentTests.swift]: Restored and updated tests

[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseVertexAI/Tests/Unit/VertexComponentTests.swift]: Thanks @andrewheard !

**Review by andrewheard — APPROVED** [review]

**Comment by paulb777** [comment]
Going to merge here. I'll address any additional comments in a future PR

File: FirebaseVertexAI/Sources/VertexAI.swift [hunk 2]
```diff
@@ -112,6 +117,9 @@ public class VertexAI: NSObject {
     )
   }
 
+  /// Class to enable VertexAI to register via the Objective-C based Firebase component system.
+  @objc(FIRVertexAIComponent) class FirebaseVertexAIComponent: NSObject {}
+
```
[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseVertexAI/Sources/VertexAI.swift:122]: No, see updated comment in #13055

**Pull request merged**